File: android/src/main/kotlin/me/pulkitagarwal/connectivity_control/NetworkInformationMapper.kt
package me.pulkitagarwal.connectivity_control.network

import android.net.NetworkCapabilities
import me.pulkitagarwal.connectivity_control.network.NetworkUtils

object NetworkInformationMapper {
    private val transports = listOf(
        NetworkCapabilities.TRANSPORT_WIFI to "wifi",
        NetworkCapabilities.TRANSPORT_VPN to "vpn",
        NetworkCapabilities.TRANSPORT_ETHERNET to "ethernet",
        NetworkCapabilities.TRANSPORT_CELLULAR to "cellular"
    )

    fun map(cap: NetworkCapabilities): List<Map<String, Any?>> {
        return transports
            .filter { cap.hasTransport(it.first) }
            .map { (_, type) -> buildNetworkMap(type, cap) }
    }

    private fun buildNetworkMap(
        type: String,
        cap: NetworkCapabilities
    ): Map<String, Any?> {
        return mapOf(
            "type" to type,
            "hasInternet" to NetworkUtils.hasInternet(cap),
            "isValidated" to NetworkUtils.isValidated(cap),
            "isMetered" to NetworkUtils.isMetered(cap),
            "downLinkKbps" to cap.linkDownstreamBandwidthKbps,
            "upLinkKbps" to cap.linkUpstreamBandwidthKbps
        )
    }
}

File: ios/connectivity_control/Sources/connectivity_control/NetworkInformationMapper.swift
import Network
import Foundation

final class NetworkInformationMapper {

    static func map(path: NWPath) -> [[String: Any?]] {
        var results: [[String: Any?]] = []

        if path.usesInterfaceType(.wifi) {
            results.append(buildNetworkMap(type: "wifi", path: path))
        }

        if path.usesInterfaceType(.cellular) {
            results.append(buildNetworkMap(type: "cellular", path: path))
        }

        if path.usesInterfaceType(.wiredEthernet) {
            results.append(buildNetworkMap(type: "ethernet", path: path))
        }

        if path.usesInterfaceType(.other) {
            results.append(buildNetworkMap(type: "other", path: path))
        }

        return results
    }

    private static func buildNetworkMap(
        type: String,
        path: NWPath
    ) -> [String: Any?] {
        return [
            "type": type,
            "hasInternet": path.status == .satisfied,
            "isValidated": path.status == .satisfied,
            "isMetered": path.isExpensive,
            "downLinkKbps": nil,
            "upLinkKbps": nil
        ]
    }
}
